### PR TITLE
sharing/handler: try serving index.html and then object listing if a …

### PR DESCRIPTION
**Pull request by jenlij**
…directory is requested

### CODE REVIEW

File: sharing/handler.go [hunk 1]
```diff
@@ -406,24 +407,41 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
+		pt = pathPrefix + pt      // e.g. "folder1/folder2/index.html"
 	}
-
-	if pt == "" {
-		pt = "index.html"
-	}
-
 	o, err := project.StatObject(ctx, bucket, pt)
-	if err != nil {
+	if err != nil && !strings.Contains(err.Error(), "object not found") {
```
Comment (egonelbre) [inline sharing/handler.go:419]: Use `errors.Is` for checking error conditions.

File: sharing/handler.go [hunk 2]
```diff
@@ -406,24 +407,41 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
 		handler.handleUplinkErr(w, "stat object", err)
 		return err
 	}
+	if err != nil && strings.Contains(err.Error(), "object not found") {
```
Comment (egonelbre) [inline sharing/handler.go:423]: Use `errors.Is` for checking error conditions.

File: sharing/handler.go [hunk 3]
```diff
@@ -406,24 +407,41 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
 		}
 	}()
 
-	// e.g. http://mydomain.com/folder2/index.html with root="bucket1/folder1"
-	rootPath := strings.SplitN(root, "/", 2) // e.g. rootPath=[bucket1, folder1]
+	// e.g. http://mydomain.com/folder2/index.html with "storj_root:bucket1/folder1/"
+	rootPath := strings.SplitN(root, "/", 2) // e.g. rootPath=[bucket1, folder1/]
```
Comment (egonelbre) [inline sharing/handler.go:411]: I think this path construction could be in some sort of helpers that figures out the exact paths. Currently it's not quite clear what's happening.

File: sharing/handler.go [hunk 4]
```diff
@@ -406,24 +407,41 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
 		return err
 	}
+	if err != nil && strings.Contains(err.Error(), "object not found") {
+		pt2 := pt + "index.html"
+		o, err = project.StatObject(ctx, bucket, pt2)
+		if err != nil && !strings.Contains(err.Error(), "object not found") {
+			handler.handleUplinkErr(w, "stat object", err)
+			return err
+		}
+		if err != nil && strings.Contains(err.Error(), "object not found") {
```
Comment (egonelbre) [inline sharing/handler.go:430]: Use errors.Is for checking error conditions.

File: sharing/handler.go [hunk 5]
```diff
@@ -406,24 +407,41 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
+			// list objects
+			serializedAccess, err := access.Serialize()
```
Comment (egonelbre) [inline sharing/handler.go:432]: It feels weird that we have to reserialize the access rather than pulling it from the cache together with the access.

**Review by egonelbre — CHANGES_REQUESTED** [review]

File: sharing/handler.go (hunk 3 above)
Comment (jenlij) [reply sharing/handler.go:411]: sure that's a good idea

File: sharing/handler.go (hunk 5 above)
Comment (jenlij) [reply sharing/handler.go:432]: Yeah I thought so too. Right now the cache only holds the unserialized access. Do you think it's worth keeping the serialized in addition?

Comment (egonelbre) [reply sharing/handler.go:432]: Sure, it shouldn't cause significant issues.

File: sharing/handler.go (hunk 1 above)
Comment (jenlij) [reply sharing/handler.go:419]: I'm having trouble with `errors.Is()` 
The error changes depending on the path requested and looks like `uplink: object not found ("<path>")` 
Is there still a way to use `errors.Is()` given this issue?

Comment (egonelbre) [reply sharing/handler.go:419]: Yes, `errors.Is` should still work.

Comment (jenlij) [reply sharing/handler.go:419]: Do you know what the target error should be? I even tried this but still no match (even though it looks like a match when printed):
```
func errwrapf(format string, err error, args ...interface{}) error {
	pe := errs.Class("uplink")
	var all []interface{}
	all = append(all, err)
	all = append(all, args...)
	return pe.Wrap(fmt.Errorf(format, all...))
}
ErrObjectNotFound := errors.New("object not found")
x :=errwrapf("%w (%q)", ErrObjectNotFound, pt)

```

Comment (egonelbre) [reply sharing/handler.go:419]: Use the errors defined in uplink package. This means `errors.Is(err, uplink.ErrObjectNotFound)`.

Comment (jenlij) [reply sharing/handler.go:419]: Ok that seems to work now, Idk what I did earlier that it wasn't working with that uplink.ErrObjectNotFound. Thanks Egon

Comment (jenlij) [reply sharing/handler.go]: done

File: sharing/handler.go (hunk 2 above)
Comment (jenlij) [reply sharing/handler.go]: done

File: sharing/handler.go (hunk 3 above)
Comment (jenlij) [reply sharing/handler.go]: done

File: sharing/handler.go (hunk 4 above)
Comment (jenlij) [reply sharing/handler.go]: done

File: sharing/handler.go (hunk 5 above)
Comment (jenlij) [reply sharing/handler.go]: done

File: sharing/handler.go (hunk 1 above)
Comment (egonelbre) [reply sharing/handler.go]: You can read more here https://blog.golang.org/go1.13-errors.

But in general `errors.Is` does not use the message of an error to compare them. So if you "redefine" an error you cannot use to compare them that way. (There are exceptions to this, but this means the specific error type needs to support it.)

File: sharing/handler.go [hunk 6]
```diff
@@ -406,25 +407,50 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
+	httpranger.ServeContent(ctx, w, r, path, o.System.Created, objectranger.New(project, o, bucket))
 	return nil
 }
+
+// determineBucketAndPath is a helper function to parse storj_root and the url into the bucket and path
+// For example, we have http://mydomain.com/folder2/index.html with storj_root:bucket1/folder1/
+// The root path will be [bucket1, folder1/]. Our bucket is named bucket1.
+// Since the url has a path of /folder2/index.html and the second half of the root path is folder1/,
+// we get an object path of folder1/folder2/index.html
+func determineBucketAndPath(root, urlPath string) (bucket, path string) {
```
Comment (egonelbre) [inline sharing/handler.go:448]: Shouldn't this be the object key rather than path?

**Review by egonelbre — COMMENTED** [review]
Looks good, but CI is complaining.

File: sharing/txtrecords.go [hunk 7]
```diff
@@ -23,29 +23,30 @@ type txtRecords struct {
 }
 
 type txtRecord struct {
-	access    *uplink.Access
-	root      string
-	timestamp time.Time
+	access           *uplink.Access
+	serializedAccess string
```
Comment (jtolio) [inline sharing/txtrecords.go:27]: wait, why?

File: sharing/handler.go [hunk 8]
```diff
@@ -406,25 +407,50 @@ func (handler *Handler) handleHostingService(ctx context.Context, w http.Respons
...
+			handler.handleUplinkErr(w, "stat object", err)
+			return err
+		}
+		if err != nil && errors.Is(err, uplink.ErrObjectNotFound) {
+			// list objects
+			if err != nil {
+				handler.log.Error("unable to handle request", zap.Error(err))
+				http.Error(w, "unable to handle request", http.StatusInternalServerError)
+			}
+			err = handler.servePrefix(ctx, w, project, serializedAccess, bucket, path)
```
Comment (jtolio) [inline sharing/handler.go:430]: could we make servePrefix take a parsed access instead?

**Review by egonelbre — APPROVED** [review]

**Review by cam-a — APPROVED** [review]

**Review by jtolio — APPROVED** [review]

**Pull request merged**